File: 40.js
function generateParenthesis(n) {
    let ans = [];

    function backtrack(cur, open, close) {
        if (cur.length === 2 * n) {
            ans.push(cur);
            return;
        }
        if (open < n) backtrack(cur + "(", open + 1, close);
        if (close < open) backtrack(cur + ")", open, close + 1);
    }

    backtrack("", 0, 0);
    return ans;
}
// Example usage:
const n = 3;   
console.log(generateParenthesis(n));

File: 40.py
def generateParenthesis(n):
    ans = []

    def backtrack(cur, open_c, close_c):
        if len(cur) == 2 * n:
            ans.append(cur)
            return
        
        if open_c < n:
            backtrack(cur + "(", open_c + 1, close_c)
        if close_c < open_c:
            backtrack(cur + ")", open_c, close_c + 1)

    backtrack("", 0, 0)
    return ans
# Example usage:
n = 5
print(generateParenthesis(n)) 